Link errors when using caliper configured with adiak and/or gotcha

My mfem configuration uses `caliper` (`MFEM_USE_CALIPER=YES`) where caliper is configured with [`adiak`](https://github.com/LLNL/Adiak) and [`gotcha`](https://github.com/LLNL/gotcha).

Since mfem is not adding `adiak` and `gotcha` on the link line, I get "undefined reference" link errors, e.g.:
```
libcaliper.a(AdiakExport.cpp.o): In function `(anonymous namespace)::export_globals_to_adiak(cali::Caliper*, cali::Channel*)':
AdiakExport.cpp:(.text+0x2b7): undefined reference to `adiak_numerical_from_type'
AdiakExport.cpp:(.text+0x2e0): undefined reference to `adiak_get_basetype'
...
libcaliper.a(wrappers.c.o): In function `curious_apply_wrappers':
wrappers.c:(.text+0x1f): undefined reference to `gotcha_wrap'
libcaliper.a(wrappers.c.o): In function `wrapped_read':
wrappers.c:(.text+0x97): undefined reference to `gotcha_get_wrappee'
...
```

I was able to get past this (in the make-based build system) by adding a few lines for the adiak and gotcha configuration, e.g.
```
ADIAK_DIR = @MFEM_DIR@/../adiak
ADIAK_OPT = -I$(ADIAK_DIR)/include
ADIAK_LIB = $(XLINKER)-rpath,$(ADIAK_DIR)/lib -L$(ADIAK_DIR)/lib -ladiak

GOTCHA_DIR = @MFEM_DIR@/../gotcha
GOTCHA_OPT = -I$(GOTCHA_DIR)/include
GOTCHA_LIB = $(XLINKER)-rpath,$(GOTCHA_DIR)/lib -L$(GOTCHA_DIR)/lib -lgotcha
```

Would it be ok to add `MFEM_USE_ADIAK` and `MFEM_USE_GOTCHA` as additional optional TPLs to mfem?

--- a/config/cmake/modules/FindCaliper.cmake
+++ b/config/cmake/modules/FindCaliper.cmake
@@ -23,21 +23,15 @@ mfem_find_package(Caliper CALIPER CALIPER_DIR
 
 # Append adiak path/lib if the user provided ADIAK_DIR
 if(ADIAK_DIR AND EXISTS ${ADIAK_DIR})
-    list(APPEND CALIPER_INCLUDE_DIRS ${ADIAK_DIR}/include)
-    
-    unset(_lib)
-    find_library(_lib "adiak" NO_CACHE PATHS ${ADIAK_DIR}/lib)
-    list(APPEND CALIPER_LIBRARIES ${_lib})
+    find_package(adiak NO_DEFAULT_PATH REQUIRED PATHS ${ADIAK_DIR}/lib/cmake/adiak ${ADIAK_DIR})
+    list(APPEND CALIPER_INCLUDE_DIRS ${adiak_INCLUDE_DIRS})
+    list(APPEND CALIPER_LIBRARIES ${adiak_LIBRARIES})
 endif()
 
 # Append gotcha path/lib if the user provided GOTCHA_DIR
 if(GOTCHA_DIR AND EXISTS ${GOTCHA_DIR})
-    list(APPEND CALIPER_INCLUDE_DIRS ${GOTCHA_DIR}/include)
-    
-    unset(_lib)
-    find_library(_lib "gotcha" NO_CACHE PATHS ${GOTCHA_DIR}/lib)
-    list(APPEND CALIPER_LIBRARIES ${_lib})
+    find_package(gotcha NO_DEFAULT_PATH REQUIRED PATHS ${GOTCHA_DIR}/lib/cmake/gotcha ${GOTCHA_DIR})
+    list(APPEND CALIPER_INCLUDE_DIRS ${gotcha_INCLUDE_DIRS})
+    list(APPEND CALIPER_LIBRARIES ${gotcha_LIBRARIES})
 endif()
-
-unset(_lib)
 

--- a/config/cmake/modules/FindCaliper.cmake
+++ b/config/cmake/modules/FindCaliper.cmake
@@ -16,7 +16,28 @@
 
 include(MfemCmakeUtilities)
 mfem_find_package(Caliper CALIPER CALIPER_DIR
-       	"include" "caliper/cali.h"
-       	"lib" "caliper"
-  	"Paths to headers required by Caliper."
-       	"Libraries required by Caliper.")
+      "include" "caliper/cali.h"
+      "lib" "caliper"
+      "Paths to headers required by Caliper."
+      "Libraries required by Caliper.")
+
+# Append adiak path/lib if the user provided ADIAK_DIR
+if(ADIAK_DIR AND EXISTS ${ADIAK_DIR})
+    list(APPEND CALIPER_INCLUDE_DIRS ${ADIAK_DIR}/include)
+    
+    unset(_lib)
+    find_library(_lib "adiak" NO_CACHE PATHS ${ADIAK_DIR}/lib)
+    list(APPEND CALIPER_LIBRARIES ${_lib})
+endif()
+
+# Append gotcha path/lib if the user provided GOTCHA_DIR
+if(GOTCHA_DIR AND EXISTS ${GOTCHA_DIR})
+    list(APPEND CALIPER_INCLUDE_DIRS ${GOTCHA_DIR}/include)
+    
+    unset(_lib)
+    find_library(_lib "gotcha" NO_CACHE PATHS ${GOTCHA_DIR}/lib)
+    list(APPEND CALIPER_LIBRARIES ${_lib})
+endif()
+
+unset(_lib)
+

--- a/config/defaults.mk
+++ b/config/defaults.mk
@@ -473,7 +473,17 @@ OCCA_LIB = $(XLINKER)-rpath,$(OCCA_DIR)/lib -L$(OCCA_DIR)/lib -locca
 # CALIPER library configuration
 CALIPER_DIR = @MFEM_DIR@/../caliper
 CALIPER_OPT = -I$(CALIPER_DIR)/include
-CALIPER_LIB = $(XLINKER)-rpath,$(CALIPER_DIR)/lib64 -L$(CALIPER_DIR)/lib64 -lcaliper
+CALIPER_LIB = $(XLINKER)-rpath,$(CALIPER_DIR)/lib -L$(CALIPER_DIR)/lib -lcaliper
+
+ifdef ADIAK_DIR
+   CALIPER_OPT += -I$(ADIAK_DIR)/include
+   CALIPER_LIB += $(XLINKER)-rpath,$(ADIAK_DIR)/lib -L$(ADIAK_DIR)/lib -ladiak
+endif
+ifdef GOTCHA_DIR
+   CALIPER_OPT += -I$(GOTCHA_DIR)/include
+   CALIPER_LIB += $(XLINKER)-rpath,$(GOTCHA_DIR)/lib -L$(GOTCHA_DIR)/lib -lgotcha
+endif
+
 
 # BLITZ library configuration
 BLITZ_DIR = @MFEM_DIR@/../blitz

--- a/config/defaults.mk
+++ b/config/defaults.mk
@@ -473,15 +473,15 @@ OCCA_LIB = $(XLINKER)-rpath,$(OCCA_DIR)/lib -L$(OCCA_DIR)/lib -locca
 # CALIPER library configuration
 CALIPER_DIR = @MFEM_DIR@/../caliper
 CALIPER_OPT = -I$(CALIPER_DIR)/include
-CALIPER_LIB = $(XLINKER)-rpath,$(CALIPER_DIR)/lib -L$(CALIPER_DIR)/lib -lcaliper
+CALIPER_LIB = $(XLINKER)-rpath,$(CALIPER_DIR)/lib64 $(XLINKER)-rpath,$(CALIPER_DIR)/lib -L$(CALIPER_DIR)/lib64 -L$(CALIPER_DIR)/lib -lcaliper
 
 ifdef ADIAK_DIR
    CALIPER_OPT += -I$(ADIAK_DIR)/include
-   CALIPER_LIB += $(XLINKER)-rpath,$(ADIAK_DIR)/lib -L$(ADIAK_DIR)/lib -ladiak
+   CALIPER_LIB += $(XLINKER)-rpath,$(ADIAK_DIR)/lib64 $(XLINKER)-rpath,$(ADIAK_DIR)/lib -L$(ADIAK_DIR)/lib64 -L$(ADIAK_DIR)/lib -ladiak
 endif
 ifdef GOTCHA_DIR
    CALIPER_OPT += -I$(GOTCHA_DIR)/include
-   CALIPER_LIB += $(XLINKER)-rpath,$(GOTCHA_DIR)/lib -L$(GOTCHA_DIR)/lib -lgotcha
+   CALIPER_LIB += $(XLINKER)-rpath,$(GOTCHA_DIR)/lib64 $(XLINKER)-rpath,$(GOTCHA_DIR)/lib -L$(GOTCHA_DIR)/lib64 -L$(GOTCHA_DIR)/lib -lgotcha
 endif
 
 

--- a/examples/caliper/CMakeLists.txt
+++ b/examples/caliper/CMakeLists.txt
@@ -33,13 +33,13 @@ add_mfem_examples(CALIPER_EXE_SRCS ${PREFIX})
 if (MFEM_ENABLE_TESTING)
   foreach(SRC_FILE ${CALIPER_EXE_SRCS})
     get_filename_component(SRC_FILENAME ${SRC_FILE} NAME)
-    string(REPLACE ".cpp" "" TEST_NAME ${SRC_FILENAME})
+    string(REPLACE ".cpp" "" TEST_NAME ${PREFIX}${SRC_FILENAME})
 
     set(THIS_TEST_OPTIONS "-no-vis")
 
     if (NOT (${TEST_NAME} MATCHES ".*p$"))
       add_test(NAME ${TEST_NAME}_ser
-        COMMAND ${TEST_NAME} ${THIS_TEST_OPTIONS})
+        COMMAND $<TARGET_FILE:${TEST_NAME}> ${THIS_TEST_OPTIONS})
     else()
       add_test(NAME ${TEST_NAME}_np=${MFEM_MPI_NP}
         COMMAND ${MPIEXEC} ${MPIEXEC_NUMPROC_FLAG} ${MFEM_MPI_NP}

